hkmc2/shared/src/test/mlscript/ucs/patterns/SimpleTuple.mls

```diff
@@ -1,78 +1,99 @@
+:js
+
 // We test the support for simple tuple patterns in this file.
 // Splice tuple patterns will be implement in the future.
 
 fun sum(x, y) = x + y
 sum(1, 2)
+//│ = 3
 
 :fixme
-fun sum'([x, y]) = x + y
-sum'([1, 2])
+fun sum([x, y]) = x + y
+sum([1, 2])
 //│ /!!!\ Uncaught error: scala.MatchError: Tup(List(Ident(x), Ident(y))) (of class hkmc2.syntax.Tree$Tup)
 
-:todo
-fun sum''(pair) =
+:ucs desugared
+fun sum(pair) =
   if pair is [x, y] then x + y
-sum''([1, 2])
-//│ ╔══[ERROR] Unrecognized pattern.
-//│ ║  l.14: 	  if pair is [x, y] then x + y
-//│ ╙──      	             ^^^^^^
-
-// We need native support for tuple patterns in MLscript syntax.
-// Otherwise the following cases work.
+sum([1, 2])
+//│ Desugared:
+//│ >  if pair@33 is [$elem0@34, $elem1@35] and
+//│ >    let x@36 = $elem0@34#0
+//│ >    let y@37 = $elem1@35#0
+//│ >    else .+#1(x@36#0, y@37#0)
+//│ = 3
 
-:todo
+:ucs desugared
 fun test(thing) =
   if thing is [] then 0
+//│ Desugared:
+//│ >  if thing@44 is [] then 0
+
+:re
 test("")
+//│ ═══[RUNTIME ERROR] Error: match error
+
+:re
 test(12)
-//│ ╔══[ERROR] Unrecognized pattern.
-//│ ║  l.25: 	  if thing is [] then 0
-//│ ╙──      	              ^^
+//│ ═══[RUNTIME ERROR] Error: match error
 
-:todo
-// Since pattern destruction is desugared to let bindings, matching with other
-// classes after the tuple pattern will not work.
+:ucs desugared
 class Point(x: Int, y: Int)
 fun discarded_cases(thing) =
   if thing is
     [x, y] then x + y
     Point(x, y) then x + y
-//│ ╔══[ERROR] Unrecognized pattern.
-//│ ║  l.38: 	    [x, y] then x + y
-//│ ╙──      	    ^^^^^^
+//│ Desugared:
+//│ >  if
+//│ >    thing@53 is [$elem0@59, $elem1@60] and
+//│ >      let x@61 = $elem0@59#0
+//│ >      let y@62 = $elem1@60#0
+//│ >      else .+#3(x@61#0, y@62#0)
+//│ >    thing@53 is Point($param0@54, $param1@55) and
+//│ >      let x@56 = $param0@54#0
+//│ >      let y@57 = $param1@55#0
+//│ >      else .+#2(x@56#0, y@57#0)
 
 :e
 :todo
 discarded_cases(Point(0, 0))
+//│ = 0
 
 // A workaround is to move the tuple pattern to the last case.
-:todo
+:ucs desugared
 fun working_cases(thing) =
   if thing is
     Point(x, y) then x + y
     [x, y] then x + y
-//│ ╔══[ERROR] Unrecognized pattern.
-//│ ║  l.53: 	    [x, y] then x + y
-//│ ╙──      	    ^^^^^^
+//│ Desugared:
+//│ >  if
+//│ >    thing@73 is Point($param0@79, $param1@80) and
+//│ >      let x@81 = $param0@79#0
+//│ >      let y@82 = $param1@80#0
+//│ >      else .+#5(x@81#0, y@82#0)
+//│ >    thing@73 is [$elem0@74, $elem1@75] and
+//│ >      let x@76 = $elem0@74#0
+//│ >      let y@77 = $elem1@75#0
+//│ >      else .+#4(x@76#0, y@77#0)
 
 working_cases(Point(0, 0))
+//│ = 0
 
 // However, the `Object` type forbids tuples to be used.
 :todo
 working_cases([0, 0])
+//│ = 0
+
 
-:todo
 fun not_working(x) =
```

Comment:
Seems like the name should be updated!

Reply:
This works right now. The support to simple tuple was limited.